util/request_source.go

```diff
@@ -73,13 +93,28 @@ func IsRequestSourceInternal(reqSrc *RequestSource) bool {
 func (r *RequestSource) GetRequestSource() string {
 	// if r.RequestSourceType is not set, it's mostly possible that r.RequestSourceInternal is not set
 	// to avoid internal requests be marked as external(default value), return unknown source here.
-	if r == nil || r.RequestSourceType == "" {
+	if r == nil || (len(r.RequestSourceType) == 0 && len(r.ExplicitRequestSourceType) == 0) {
 		return SourceUnknown
 	}
+	clearEmpty := func(list []string) []string {
+		i := 0
+		for _, v := range list {
+			if len(v) != 0 {
+				list[i] = v
+				i++
+			}
+		}
+		return list[:i]
+	}
 	if r.RequestSourceInternal {
-		return InternalRequest + r.RequestSourceType
+		labels := []string{InternalRequest, r.RequestSourceType, r.ExplicitRequestSourceType}
+		labels = clearEmpty(labels)
+
+		return strings.Join(labels, "_")
 	}
-	return ExternalRequest + r.RequestSourceType
+	labels := []string{ExternalRequest, r.RequestSourceType, r.ExplicitRequestSourceType}
```

Comment:
Currently, tikv uses the last segments of `resquest_source.split("_")` as the explicit_request_source(task_names) value, so if it is empty, the logic won't be correct. Because there are some request_source_types such as `ddl_refill`, we can't depend on the number of segments either. So I suggest always make ExplicitRequestSourceType not empty.

Reply:
changed,ptal @glorv